pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -411,13 +411,18 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
 // AssignedPodAdded is called when a bound pod is added. Creation of this pod
 // may make pending pods with matching affinity terms schedulable.
 func (p *PriorityQueue) AssignedPodAdded(pod *v1.Pod) {
+	p.lock.Lock()
 	p.movePodsToActiveQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod))
+	p.lock.Unlock()
 }
 
 // AssignedPodUpdated is called when a bound pod is updated. Change of labels
 // may make pending pods with matching affinity terms schedulable.
 func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
+	p.lock.Lock()
+	p.deleteNominatedPodIfExists(pod)
```

Comment:
The reason why I didn't put the logic in `assume()` is current assume design has no interaction with scheduling_queue, if we have to do that, it will result in pretty a lot of interface changes.

I don't quite understand "double counting". Could you collaborate a bit?

Based on current solution, in between assume() and bound(), the nominatedMap is still kept, so it's still calculated when scheduling low priority pending pods. But the interval would be pretty short.

Reply:
"double counting" comes from the fact that an assumed pod is added to the NodeInfo in the scheduler cache and its resources are deducted from the node. It also remains in the "NominatedNode" map. So, when scheduler tries scheduling another pod, it adds the resources needed by the "nominated pods" again to the resources used on the node. So, it may determine the incoming pod unschedulable due to lack of resources, while there may be enough resources.

The easiest way to fix it is probably adding a pointer of the scheduling queue to `factory.Config`. It will then be available to the `scheduler.Scheduler`. Then you can expose and call `deleteNominatedPodIfExists` form `scheduler.go -> assum()`.

Reply:
Ah I see the point. Thanks for the explanation.

Will put logic into assume().